etk-asm/src/ingest.rs

```diff
@@ -403,70 +283,87 @@ where
     where
         P: Into<PathBuf>,
     {
-        let nodes = parse_asm(src)?;
-        let partial = self.sources.resolve(path.into(), Scope::independent())?;
-        partial.push(nodes);
-
-        while let Some(source) = self.sources.peek() {
-            let node = match source.nodes.next() {
-                Some(n) => n,
-                None => {
-                    self.sources.pop()?;
-                    continue;
-                }
-            };
+        let mut program = Program::new(path.into());
+        let nodes = self.preprocess(&mut program, text)?;
+        let mut asm = Assembler::new();
+        let raw = asm.assemble(nodes)?;
+
+        self.output.write_all(&raw).context(error::Io {
+            message: "writing output",
+            path: None,
+        })?;
 
+        Ok(())
+    }
+
+    fn preprocess(&mut self, program: &mut Program, src: &str) -> Result<Vec<RawOp>, Error> {
+        let nodes = parse_asm(src).context(error::Parse {
+            path: program.sources.last().unwrap().clone(),
+        })?;
+        let mut raws = Vec::new();
+        for node in nodes {
             match node {
                 Node::Op(op) => {
-                    self.sources.write(RawOp::Op(op))?;
+                    raws.push(RawOp::Op(op));
                 }
-                Node::Raw(raw) => {
-                    self.sources.write(RawOp::Raw(raw))?;
+                Node::Import(imp_path) => {
+                    let new_raws = self.resolve_and_ingest(program, imp_path)?;
+                    raws.extend(new_raws);
                 }
-                Node::Import(path) => {
-                    let partial = self.sources.resolve(path, Scope::same())?;
-                    let parsed = parse_file(partial.path())?;
-                    partial.push(parsed);
+                Node::Include(inc_path) => {
+                    let inc_raws = self.resolve_and_ingest(program, inc_path)?;
+                    raws.push(RawOp::Scope(inc_raws));
                 }
-                Node::Include(path) => {
-                    let partial = self.sources.resolve(path, Scope::independent())?;
-                    let parsed = parse_file(partial.path())?;
-                    partial.push(parsed);
-                }
-                Node::IncludeHex(path) => {
-                    let partial = self.sources.resolve(path, Scope::same())?;
-
+                Node::IncludeHex(hex_path) => {
                     let file =
-                        std::fs::read_to_string(partial.path()).with_context(|_| error::Io {
-                            message: "reading hex include",
-                            path: partial.path().to_owned(),
+                        std::fs::read_to_string::<&PathBuf>(&hex_path).with_context(|_| {
+                            error::Io {
+                                message: "reading hex include",
+                                path: <PathBuf as Into<PathBuf>>::into(hex_path.clone()).to_owned(),
```

Comment:
This seems a bit crazy. Can't just do:

```suggestion
                                path: hex_path,
```

or maybe:

```suggestion
                                path: &hex_path,
```

?

Reply:
This is rare. This was, after quite a while, the only way I could get the code to work. But I just removed it and it just works :open_mouth:

Reply:
snafu sometimes amazes me, and sometimes confounds me.